source/common/http/conn_pool_grid.h

```diff
@@ -24,45 +24,63 @@ class ConnectivityGrid : public ConnectionPool::Instance {
-                       absl::string_view transport_failure_reason,
-                       Upstream::HostDescriptionConstSharedPtr host) override;
-    void onPoolReady(RequestEncoder& encoder, Upstream::HostDescriptionConstSharedPtr host,
-                     const StreamInfo::StreamInfo& info,
-                     absl::optional<Http::Protocol> protocol) override;
+    // This holds state for a single connection attempt to a specific pool.
+    struct ConnectionAttemptState : public ConnectionPool::Callbacks,
+                                    public LinkedObject<ConnectionAttemptState> {
+      ConnectionAttemptState(WrapperCallbacks& parent, PoolIterator it);
+
+      // ConnectionPool::Callbacks
+      void onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                         absl::string_view transport_failure_reason,
+                         Upstream::HostDescriptionConstSharedPtr host) override;
+      void onPoolReady(RequestEncoder& encoder, Upstream::HostDescriptionConstSharedPtr host,
+                       const StreamInfo::StreamInfo& info,
+                       absl::optional<Http::Protocol> protocol) override;
+
+      ConnectionPool::Instance& pool() { return **pool_it_; }
+
+      // A pointer back up to the parent.
+      WrapperCallbacks& parent_;
+      // The pool for this connection attempt.
+      const PoolIterator pool_it_;
+      // The handle to cancel this connection attempt.
+      // This is owned by the pool which created it.
+      Cancellable* cancellable_;
+    };
+    using ConnectionAttemptStatePtr = std::unique_ptr<ConnectionAttemptState>;
 
     // ConnectionPool::Cancellable
     void cancel(Envoy::ConnectionPool::CancelPolicy cancel_policy) override;
 
     // Removes this from the owning list, deleting it.
     void deleteThis();
 
-    ConnectionPool::Instance& pool() { return **pool_it_; }
+    // Called on pool failure or timeout to kick off another connection attempt.
+    // Returns true if there is a failover pool and a connection has been
+    // attempted, false if all pools have been tried.
+    bool tryAnotherConnection();
+
+    // Tracks all the connection attempts which currently in flight.
+    std::list<ConnectionAttemptStatePtr> connection_attempts_;
+
     // The owning grid.
     ConnectivityGrid& grid_;
     // The decoder for the original newStream, needed to create streams on subsequent pools.
     Http::ResponseDecoder& decoder_;
     // The callbacks from the original caller, which must get onPoolFailure or
     // onPoolReady unless there is call to cancel().
     ConnectionPool::Callbacks& inner_callbacks_;
-    // The current pool being connected to.
-    PoolIterator pool_it_;
-    // The handle to cancel the request to the current pool.
-    // This is owned by the pool which created it.
-    Cancellable* cancellable_;
+    // The timer which tracks when new connections should be attempted.
+    Event::TimerPtr failover_timer_;
```

Comment:
I would prefer to avoid the term failover here - it implies that we've given up on the first attempt and moved to the second. Here if the timeout fires, we should start the next attempt but keep running this one until one of them succeeds. How about `next_attempt_timer_`?